apps/api/src/tests/services/application.ts

```diff
@@ -0,0 +1,197 @@
+ */
+
+import assert from 'node:assert';
+import { after, before, describe, it } from 'node:test';
+
+import { PostgreSqlContainer, StartedPostgreSqlContainer } from '@testcontainers/postgresql';
+
+import { ApplicationStates } from 'pcgl-daco/packages/data-model/src/types.ts';
+import { initMigration, startDb, type PostgresDb } from '../../db/index.ts';
+import { applications } from '../../db/schemas/applications.ts';
+import service from '../../service/application-service.ts';
+
+const PG_DATABASE = process.env.PG_DATABASE || 'testUser';
+const PG_USER = process.env.PG_USER || 'testPassword';
+const PG_PASSWORD = process.env.PG_PASSWORD || 'postgres';
+
+describe('Application Service', () => {
+	let db: PostgresDb;
+	let applicationService: ReturnType<typeof service>;
+	let container: StartedPostgreSqlContainer;
+
+	const user_id = 'testUser@oicr.on.ca';
+
+	before(async () => {
+		container = await new PostgreSqlContainer()
+			.withUsername(PG_USER)
+			.withPassword(PG_PASSWORD)
+			.withDatabase(PG_DATABASE)
+			.start();
+
+		const connectionString = container.getConnectionUri();
+		db = startDb(connectionString);
+
+		await initMigration(db);
+
+		applicationService = service(db);
+		// TODO: create file with seed data for postgres to test read actions
+		// Not yet supported in Drizzle Kit:
+		// https://orm.drizzle.team/docs/kit-seed-data
+	});
+
+	it('should create applications with status DRAFT and submitted user_id', async () => {
+		const application = await applicationService.createApplication({ user_id });
+
+		assert.notEqual(application, null);
+		assert.strictEqual(application?.user_id, user_id);
+		assert.strictEqual(application?.state, ApplicationStates.DRAFT);
+	});
+
+	it('should get applications requested by id, with application_contents', async () => {
+		const applicationRecords = await applicationService.listApplications({ user_id });
+
+		assert.ok(Array.isArray(applicationRecords));
+
+		const { id } = applicationRecords[0];
+
+		const requestedApplication = await applicationService.getApplicationById({ id });
+
+		assert.notEqual(requestedApplication, null);
+		assert.strictEqual(requestedApplication?.id, id);
+		assert.strictEqual(requestedApplication?.id, requestedApplication?.contents?.application_id);
+	});
+
+	it('should list applications filtered by user_id', async () => {
+		await applicationService.createApplication({ user_id });
+		await applicationService.createApplication({ user_id });
+
+		const applicationRecords = await applicationService.listApplications({ user_id });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+	});
+
+	it('should list applications filtered by state', async () => {
+		const applicationRecords = await applicationService.listApplications({ state: ApplicationStates.DRAFT });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+	});
```

Comment:
How do you initialize the data for this test? It looks like it depends on previous testa and that if I run this test directly it will fail. We want our unit tests to be repeatable and standalone, able to run in any order or even on their own.

If there is a series of steps that we can't separate to test independently, and would be valuable to test in sequence, then we should build these as a single test with sub tests. That way we are assured to run them in sequence.

Reply:
This comment applies to many tests in this file. We should have an script that inserts a test data set before running the tests that read data. We can use our create application api to insert these records.

Reply:
See below comment -- I'm aware the test setup could use improvements and felt that was one piece worth revisiting. I should have expanded on my notes so you were aware too!

Reply:
Added test-data init functions to testUtils